pkg/jobs/jobfrontier/frontier_test.go

```diff
@@ -464,3 +465,111 @@ func TestGetAllResolvedSpans(t *testing.T) {
 		})
 	}
 }
+
+func TestRandomizedFrontier(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	srv := serverutils.StartServerOnly(t, base.TestServerArgs{})
+	defer srv.Stopper().Stop(ctx)
+
+	s := srv.ApplicationLayer()
+
+	// Seed for reproducible randomness
+	rng, _ := randutil.NewTestRand()
+
+	// Create 50 adjacent key spans
+	const numSpans = 50
+	spans := make([]roachpb.Span, numSpans)
+	for i := 0; i < numSpans; i++ {
+		startKey := []byte{byte(i)}
+		endKey := []byte{byte(i + 1)}
+		spans[i] = roachpb.Span{Key: startKey, EndKey: endKey}
+	}
+
+	frontier, err := span.MakeFrontier(spans...)
+	require.NoError(t, err)
+
+	// Initialize spans to random timestamps 0-3
+	initialTimestamps := make([]hlc.Timestamp, numSpans)
+	for i, sp := range spans {
+		ts := hlc.Timestamp{WallTime: int64(rng.Intn(4))}
+		initialTimestamps[i] = ts
+		_, err := frontier.Forward(sp, ts)
+		require.NoError(t, err)
+	}
+
+	jobID := jobspb.JobID(3000)
+	spanSize := spans[0].Size()
+	smallChunkSize := spanSize * rng.Intn(10)
+	largeChunkSize := spanSize * numSpans * 2
+
+	persist := func() {
+		require.NoError(t, s.InternalDB().(isql.DB).Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			// Store sharded version (small chunk size forces sharding)
+			err := storeChunked(ctx, txn, jobID, "sharded", frontier, smallChunkSize)
+			require.NoError(t, err)
+
+			// Store unsharded version (large chunk size prevents sharding)
+			err = storeChunked(ctx, txn, jobID, "unsharded", frontier, largeChunkSize)
+			require.NoError(t, err)
+
+			return nil
+		}))
+	}
+
+	getAndVerify := func() {
+		require.NoError(t, s.InternalDB().(isql.DB).Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+
+			shardedFrontier, found, err := Get(ctx, txn, jobID, "sharded")
+			require.NoError(t, err)
+			require.True(t, found, "updated sharded frontier should be found")
+
+			checkSpanFrontierEquality(t, frontier, shardedFrontier)
+
+			unshardedFrontier, found, err := Get(ctx, txn, jobID, "unsharded")
+			require.NoError(t, err)
+			require.True(t, found, "updated unsharded frontier should be found")
+
+			checkSpanFrontierEquality(t, frontier, unshardedFrontier)
+
+			return nil
+		}))
+	}
+
+	persist()
+	getAndVerify()
+
+	// Randomly update some spans in the in-memory frontier
+	numUpdates := rng.Intn(20) + 10
+
+	for i := 0; i < numUpdates; i++ {
+		spanIdx := rng.Intn(numSpans)
+		newTs := hlc.Timestamp{WallTime: int64(rng.Intn(4))}
+
+		// Only update if new timestamp is greater (frontiers only move forward)
+		if initialTimestamps[spanIdx].Less(newTs) {
```

Comment:
You technically don't need to do this check since `Forward` will no-op if you're not forwarding to a later time